implement VectorTools::compute_global_error

tjhei opened the pull request:
add a function that computes global errors from cellwise errors obtained
by VectorTools::integrate_difference(). Do MPI collective if necessary.

tjhei commented:
If somebody has a better suggestion for a name for this function, please go ahead and tell me.

davydden commented:
perhaps something along the lines of `global_error_norm` or `total_error_norm`?

bangerth commented on include/deal.II/numerics/vector_tools.h:
```diff
@@ -1845,6 +1845,10 @@ namespace VectorTools
    * exact_solution matches that of the finite element used by @p dof.
    *
+   * To compute a global error norm of a finite element solution, use
+   * error_norm_from_cell_errors() with the Vector computed with this
```
maybe qualify the name with the namespace, to make it even clearer?

Vector -> output vector

bangerth commented on include/deal.II/numerics/vector_tools.h:
```diff
@@ -1904,38 +1908,8 @@ namespace VectorTools
-   * instead, and similar considerations hold when computing the $L_\infty$
-   * norm of the error.
+   * from different processors need to be combined, see
+   * error_norm_from_cell_errors().
```
same here.

bangerth commented on include/deal.II/numerics/vector_tools.h:
```diff
@@ -1996,6 +1970,32 @@ namespace VectorTools
+   * <tt>tria.n_active_cells()</tt> entries and returns the global
+   * error as given by @p norm.
+   *
+   * The @p cellwise_error Vector is typically an output given by
```
Vector -> vector

bangerth replied:
output given -> output produced

bangerth commented on include/deal.II/numerics/vector_tools.h:
```diff
@@ -1996,6 +1970,32 @@ namespace VectorTools
+   * error as given by @p norm.
+   *
+   * The @p cellwise_error Vector is typically an output given by
+   * integrate_difference and you normally want to supply the same value
```
namespace qualification

bangerth commented on include/deal.II/numerics/vector_tools.templates.h:
```diff
@@ -6662,7 +6662,57 @@ namespace VectorTools
+      case H1_norm:
+      case Hdiv_seminorm:
+      {
+        double local = cellwise_error.l2_norm();
```
const

bangerth replied:
same below

bangerth commented on include/deal.II/numerics/vector_tools.templates.h:
```diff
@@ -6662,7 +6662,57 @@ namespace VectorTools
+        return Utilities::MPI::max (local, comm);
+      }
 
+      /* TODO:
```
still to do?

davydden replied:
if we agree on the comment above, one can remove `TODO` and the commented block.

tjhei replied:
yes...

bangerth commented:
I like @davydden 's suggestion with `global_`.

davydden commented on include/deal.II/numerics/vector_tools.templates.h:
```diff
@@ -6662,7 +6662,57 @@ namespace VectorTools
+    switch (norm)
+      {
+      case L2_norm:
+      case H1_seminorm:
```
do these `H1_seminorm`, `H1_norm`, `Hdiv_seminorm` actually make sense in this context? We deal with scalars, not functions here. I would suggest to raise an error and only allow `L2_norm`, `L1_norm`, `Linfty_norm`. I don't think anyone will use `Lp_norm` in this context.

bangerth replied:
Definitely makes sense. In order to compute the *global* H1 norm, you need to compute the H1 norm on every cell and then add up their squares over all cells.

davydden replied:
Ok, i see what you mean. But I was considering this function to operate on errors without the underlying interpretation of where these errors come from. Whereas you and Timo think of it as a companion to `integrate_difference()`. 

I agree that in this context it makes sense and the documentation reflects their marriage already.

tjhei replied:
@davydden , yes otherwise it would be little-l2, etc.

tjhei commented:
Maybe ``compute_global_error_from_cellwise_errors``?

tjhei commented:
To be able to do the Lp-norm, I would need an additional parameter for the exponent similar to integrate_difference. Is it worth adding now? Do we care?

tjhei commented:
> Maybe compute_global_error_from_cellwise_errors?

Maybe ``compute_global_error``?

davydden commented:
> Maybe compute_global_error?

i like that. `compute_global_error_from_cellwise_errors` was a bit long.

bangerth commented:
I'm not concerned about long names -- the IDEs expand that just fine. I don't care too much about Lp norms.

tjhei commented:
> I don't care too much about Lp norms.

It seems nobody cares but there is not a single test for them. :-(

I am working on a revision and let y'all know when it is ready.

tjhei commented:
Okay, updated. We should probably use the function in every example where we use ``integrate_difference``.

tjhei commented:
or do you see some value in teaching that ``.l2_norm()`` is the right thing to add up cellwise L2 errors?

bangerth commented:
No value in that. We should use the function.

tjhei commented:
do not merge yet, more stuff coming!

tjhei commented:
Okay, I added a shared::Tria test, updated all examples, squashed the commits, etc..

davydden commented:
Looks good to me :+1:

tjhei commented:
Great. I think this is ready to be merged.

davydden commented:
Since @bangerth had way more comments, i leave it to him :smile:

tjhei commented:
I added a bunch of documentation, see here for a preview: http://www.math.clemson.edu/~heister/dealii-doc-temp/doxygen/deal.II/namespaceVectorTools.html#a69967cb7a148a7169963126249213db1

It would be great if somebody could proof-read.

davydden commented:
(i) I think in other places we use `K` to denote the cell/element. I saw that at least in tutorials. Maybe change `T` to `K` and `k` (component) to something else (`m`?). Also `K` is kind of conventional way to denote an element/cell.

(ii) I would reserve `\cdot` only for tensor scalar product and not use it for scalars like the very first definition (if I read it correct). One can add space by `\,`, if needed.

(iii) something is wrong with `Lp_norm` equations (latex part).

(iv) i think something is wrong with `Hdiv_seminorm` (math side). Appears to be a divergence of a scalar.

(v) common to all global expressions, you are missing the weight function there.

Other suggestions:

> where component $k$ is called $f_k$ 

where component $k$ is denoted by $f_k$

> let $w_k$ be the weight

 Let $w:\Omega \rightarrow \mathbb{R}^k$ be the weight

> Then the following possibilities are implemented, where $E_T$ is the scalar on cell $T$ computed by integrate_difference() and $E$ is the scalar computed by compute_global_error(). 

In what follows by $E_K$ we denote the local error computed by integrate_difference() whereas $E$ is the global error computed by copute_global_error().

> The absolute value to the pth power

The absolute value to the $p$-th power

tjhei commented:
> (i) I think in other places we use K to denote the cell/element. I saw that at least in tutorials. Maybe change T to K and k (component) to something else (m?). Also K is kind of conventional way to denote an element/cell.

fixed.

> (ii) I would reserve \cdot only for tensor scalar product and not use it for scalars like the very first definition (if I read it correct). One can add space by \,, if needed.

okay, I only left it when I take the product ``e \cdot w`` (two vectors).

> (iii) something is wrong with Lp_norm equations (latex part).

fixed.

> (iv) i think something is wrong with Hdiv_seminorm (math side). Appears to be a divergence of a scalar.

It looks correct to me. Note that it is ``\nabla \cdot e`` not ``\nabla \cdot e_k``.

> (v) common to all global expressions, you are missing the weight function there.

Yeah, I would hope nobody would notice. :-) The problem is that there is no good way to write a nice expression like ``\| e \|_{L^p}`` with an arbitrary weight. This is because the weight is multiplied outside of derivatives/local norms, so ``e \cdot w`` is wrong in most cases. I really hate removing expressions like ``\| \nabla \cdot e \|_{L^2}``. :-( Suggestions? Add a sentence like "The last expression assumes w=1"?

> Other suggestions:

Done, thanks! (html will be updated in a minute)

davydden commented:
> > (iv) i think something is wrong with Hdiv_seminorm (math side). Appears to be a divergence of a scalar.

> It looks correct to me. Note that it is \nabla \cdot e not \nabla \cdot e_k.

The way it visualizes at the moment is `\nabla \cdot (e \cdot w)`, `e \cdot w` is scalar, so `\nabla \cdot <scalar>` (divergence of a scalar) does not make sense.

tjhei commented:
> The way it visualizes at the moment is \nabla \cdot (e \cdot w), e \cdot w is scalar, so \nabla \cdot <scalar> (divergence of a scalar) does not make sense.

Oh, I don't mean dot product between e and w.

davydden commented:
> > (v) common to all global expressions, you are missing the weight function there.

> Add a sentence like "The last expression assumes w=1"?

maybe something like `whatever-here-with-w = \{ w \equiv 1 \} = || e ||_2`. At least this is the notation I was tough.

davydden commented:
> Oh, I don't mean dot product between e and w.

then please remove it :smile:

tjhei commented:
> then please remove it

is it clear that ``f w`` means ``(f_i w_i)_i`` (component by component scaling)?

davydden commented:
with `Hdiv_seminorm` i suppose you mean that `w` actually have to be a scalar. So it is `\nabla \cdot (w e)` and makes sense only for vector-valued `e`. But then you can't use the same `w` as above as it denotes vector valued weight. Maybe `w_0`.

tjhei commented:
> with Hdiv_seminorm i suppose you mean that w actually have to be a scalar.

No. ``f_c`` is scaled by ``w_c``.

davydden commented:
> is it clear that f w means (f_i w_i)_i (component by component scaling)?

i would say no. If `f` and `w` are tensors of first order, then it should be either tensor product `t \otimes w` or scalar product `f \cdot w`. I would suggest to write it in component notation to avoid confusing, i.e.

`\nabla_i (f_i w_i)` is probably what you want.

davydden commented:
in tensor notation that would be `\trace (\nabla \cdot ( f \otimes w) )`, if I am not mistaken.

tjhei commented:
> in tensor notation that would be \trace (\nabla \cdot ( f \otimes w) ), if I am not mistaken.

or ``\nabla cdot tr (f \otimes w)``. But I don't like that either. :-(

davydden commented:
> or \nabla cdot tr (f \otimes w).

this does not look right to me. Trace of a second order tensor `tr(A) = A_{ii}`, so `tr(f \otimes w) \equiv f \cdot w`.

What I wrote originally 

> \rm{trace} (\nabla \cdot ( f \otimes w) ),

is also wrong :smile:  as it `\nabla \cdot ( f \otimes w)` is a first order tensor. So i would stick with index notation for a moment.

tjhei commented:
okay, I went through everything one more time and fixed all the cases where the weight multiplication was confusing.

I also realized that ``W1infty_norm`` won't work for ``compute_global_norm`` and asserted/documented that.

tjhei commented:
Okay, one more formula fixed. I hope this is it now.

davydden commented:
is the documentation preview updated?

tjhei commented:
> is the documentation preview updated?

yes!

davydden commented:
It looks good to me but i will let @bangerth to have a quick look at the formulas as well before merging.

p.s. looking at the expressions in a mixture of tensor and index notation, one realises that it would not straight forward to write it using purely tensor notation.

tjhei commented:
> p.s. looking at the expressions in a mixture of tensor and index notation, one realises that it would not straight forward to write it using purely tensor notation.

I think it is much nicer this way anyways (ignoring the weights for the compact expression at the bottom). Most people will only use the weights to select components (if at all). I call this a win-win. ;-)

tjhei commented:
> It looks good to me

Your help with this is very much appreciated! I understand why nobody had written down the formulas in the first place. So much work...

bangerth commented on examples/step-48/step-48.cc:
```diff
@@ -467,7 +467,9 @@ namespace Step48
-      std::sqrt(Utilities::MPI::sum (norm_per_cell.norm_sqr(), MPI_COMM_WORLD));
+      VectorTools::compute_global_error(triangulation,
+                                        norm_per_cell,
+                                        VectorTools::L2_norm);
```
yes, neat. So much nicer :-)

bangerth commented on include/deal.II/numerics/vector_tools.h:
```diff
@@ -317,61 +319,238 @@ namespace VectorTools
+   * computed by compute_global_error(). Note that integrals are
+   * approximated by quadrature in the usual way:
+   * @f[
+   * \int_A f(x) dx \approx \sum_q f(q) \omega_q.
```
it should be `f(x_q)` where `x_q` is the q-th quadrature point mapped from the reference cell to the domain A.

bangerth commented on include/deal.II/numerics/vector_tools.h:
```diff
@@ -317,61 +319,238 @@ namespace VectorTools
+     * the mean of a function by a factor of $\frac{1}{|\Omega|}$. To
+     * compute the mean you can also use compute_mean_value(). Finally,
+     * pay attention to the sign: if $\hat{f}=0$, this will compute the
+     * negative of the mean of $f$.
```
aw, really? who implemented it that way!?

tjhei replied:
It makes sense for ``integrate_difference``, because dividing by the local area is probably not what you want. I didn't want to make a special case for ``compute_global_error``, but I guess we could still change this. Thoughts?

bangerth replied:
The average doesn't bother me. It's that we compute the average of `$\hat f - f$`, not the other way around.

bangerth commented on include/deal.II/numerics/vector_tools.h:
```diff
@@ -317,61 +319,238 @@ namespace VectorTools
+     * or, for $w \equiv 1$:
+     * @f[
+     *   E  = \sup_\Omega \|e\|_\infty = \| e \|_{L^\infty}.
+     * @f]
```
Can you add a sentence that says that in practice the maximum is not computed as the `\sup`, but as the `\max` over all quadrature points?

(I'm thinking whether there is a point of using `\sup` instead of `\max` in the definition above, but the distinction of course doesn't matter when approximated on a finite set of quadrature points.)

bangerth replied:
By the way, I say this because of the fact that you only talk about quadrature above in the context of *integrals*.

bangerth commented on include/deal.II/numerics/vector_tools.h:
```diff
@@ -317,61 +319,238 @@ namespace VectorTools
     /**
-     * #L2_norm of the divergence of a vector field
+     * #L2_norm of the divergence of a vector field. This assumes $f$ has
+     * $k = \text{dim}$ components:
```
is it *exactly* dim components, or do we just take the first dim components?

tjhei replied:
The first dim are used, my bad. Maybe we should compute the divergence over all components and let the user select the dim components she wants by using the weights? That would allow us to compute the divergence even if they are not the first dim elements.

bangerth replied:
You can't compute the divergence of *all* components :-)

I think the function does the best it can. I'm ok with its current operation, but we should document/clarify the issue with *at least* dim components in the sentence above.

tjhei replied:
> You can't compute the divergence of all components :-)

Yes, one would need to check that exactly ``dim`` weights are non-zero and use the x derivative of the component with the first non-zero weight, etc.. Whatever.

 

> we should document/clarify the issue

Yep, I did that.

bangerth commented on include/deal.II/numerics/vector_tools.h:
```diff
@@ -317,61 +319,238 @@ namespace VectorTools
+     * or, for $w \equiv 1$:
+     * @f[
+     *   E = \| \nabla e \|_{L^\infty}.
+     * @f]
```
same here about the approximation on quadrature points.

bangerth commented on include/deal.II/numerics/vector_tools.h:
```diff
@@ -317,61 +319,238 @@ namespace VectorTools
+     * because it is impossible to compute the sum of the global
+     * norms from the values $E_K$. As a work-around, you can compute the
+     * global #Linfty_norm and #W1infty_seminorm separately and then add them
+     * to get (with $w \equiv 1$):
```
shouldn't this also be true about the H1 (full) norm?

bangerth replied:
Ah, scratch that.

bangerth commented on include/deal.II/numerics/vector_tools.h:
```diff
@@ -1996,6 +2150,37 @@ namespace VectorTools
                              const double exponent = 2.);
 
   /**
+   * Takes a Vector @p cellwise_error of errors on each cell with
```
takes -> take

bangerth commented on include/deal.II/numerics/vector_tools.h:
```diff
@@ -1996,6 +2150,37 @@ namespace VectorTools
 
   /**
+   * Takes a Vector @p cellwise_error of errors on each cell with
+   * <tt>tria.n_active_cells()</tt> entries and returns the global
```
returns -> return

bangerth replied:
(or say "This function TAKES ... and RETURNS...")

bangerth commented:
Nice job -- this was probably a LOT more work than you bargained for.

Please fix the last few comments I had and merge yourself.

(As an aside, one could alternatively have let the existing `integrate_difference` function return the *global* norm as a return value -- it currently returns nothing. This would have avoided the awkwardness of having to pass the triangulation into the `compute_global_error` function, which needs it only for the communication, and the need to pass matching norm arguments to the two functions. But your solution is just as good. It also avoids the parallel communication the user may not actually want.)

tjhei commented:
> Nice job -- this was probably a LOT more work than you bargained for.

Yes, and most time was spent in constructing the testcases. But I remember going through the code before because I was unsure of some of the definitions used. Especially the handling of the weights is not intuitive.

> (As an aside, one could alternatively have let the existing integrate_difference function return the global norm as a return value -- it currently returns nothing

Neat, I didn't think of this. We could still change it...

bangerth commented:
I'll leave it up to you whether you want to integrate this into the existing function. I'm generally not too fond of functions that return things both through an argument and as a return value, as I find that confusing. Since you already have the separate function ready, let's leave it as you have it.

(Since we now have move semantics, I've always thought that we should change many of the functions in `VectorTools` and elsewhere to actually *return* what they compute, rather than use an output argument. I think that's more in the spirit of C++. But it would preclude us returning both a vector and a scalar as I had suggested above.)

tjhei commented:
> Since you already have the separate function ready, let's leave it as you have it.

Good point. I am leaving things as is.
